network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use channel::libra_channel;
+use futures::{sink::SinkExt, StreamExt};
+use libra_config::config::RoleType;
+use libra_logger::prelude::*;
+use libra_metrics::{register_histogram, DurationHistogram};
+use libra_network_address::NetworkAddress;
+use libra_types::on_chain_config::{OnChainConfigPayload, ValidatorSet};
+use network::{
+    common::NetworkPublicKeys,
+    connectivity_manager::{ConnectivityRequest, DiscoverySource},
+};
+use once_cell::sync::Lazy;
+use std::{convert::TryFrom, time::Instant};
+
+/// Histogram of idle time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_IDLE_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_idle_duration_s",
+            "Histogram of idle time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+/// Histogram of busy time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_BUSY_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_busy_duration_s",
+            "Histogram of busy time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+/// Listener which converts published  updates from the OnChainConfig to ConnectivityRequests
+/// for the ConnectivityManager.
+pub struct ConfigurationChangeListener {
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    role: RoleType,
+}
+
+/// Extracts a set of ConnectivityRequests from a ValidatorSet which are appropriate for a Validator Network.
+fn extract_validator_updates(validator_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let validator_keys = validator_set.payload().to_vec();
+
+    let mut updates = Vec::new();
+
+    // Collect the set of address updates.
+    let address_map = validator_keys
+        .iter()
+        .filter(|validator| {
+            match NetworkAddress::try_from(&validator.config().validator_network_address) {
+                Ok(_) => true,
+                Err(e) => {
+                    warn!(
+                        "Cannot parse validator network address from validator info: {}",
+                        e
+                    );
+                    false
+                }
+            }
+        })
+        .map(|validator| {
```

Comment:
I think you can squash this two closure into one with flat_map https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map
```
flat_map(|validator| {
  match NetworkAddress::try_from(&validator.config().validator_network_address) {
    Ok(addr) => Some((*validator.account_address(), addr),
    Err(e) => {
       warn!();
       None
    }
  }
}).collect()
```

Reply:
done.